### DISCUSSION TITLE
Searching for places in codebase that deal with newline characters

### DISCUSSION THREAD

**Discussion by egwetzel** [pytest-dev/pytest#10960] [Q&A]
Hi! When assertion errors appear, they include the \n character for newlines in the assertion statement in the pytest output, even when that character is not explicitly typed in code. Could someone point me to the area of the codebase where this newline character is accounted for so that it is included with the output? I'm trying to find an entry point to work on #10704 with @HSamiul. I've looked into a couple different python documents, specifically util.py, but I'm struggling to figure out exactly where the output for the user is determined and assigned.

Thanks for the help!

**Comment by portyu9** [comment]
Hi @egwetzel 

For current pytest, the main place to look is:

```text
src/_pytest/assertion/compare_text.py
```

In particular, `_diff_text()` and the surrounding string-comparison functions are where the detailed string diff for failures such as issue #10704 is generated.

The dispatch from a generic equality comparison eventually reaches the text-comparison implementation when both operands are strings.

One important distinction: the `\n` you see in the short representation is generally coming from the **representation of the string** (`repr`-style output). The problem in #10704 is that the expanded character-by-character diff displays the actual whitespace, making a newline/space difference hard to see.

So if you are working specifically on #10704 today, I would start in:

```text
src/_pytest/assertion/compare_text.py
```

rather than `util.py`. The assertion code has been reorganized since this discussion was opened.

Please mark this answer as accepted if it helped, thank you.
